butane_core/src/codegen/dbobj.rs

```diff
@@ -29,12 +29,40 @@ pub fn impl_dbobject(ast_struct: &ItemStruct, config: &Config) -> TokenStream2 {
     let pklit = make_ident_literal_str(&pkident);
     let auto_pk = is_auto(&pk_field);
 
+    let values: Vec<TokenStream2> = push_values(ast_struct, |_| true);
     let insert_cols = columns(ast_struct, |f| !is_auto(f));
-    let save_cols = columns(ast_struct, |f| !is_auto(f) && f != &pk_field);
 
-    let mut post_insert: Vec<TokenStream2> = Vec::new();
-    add_post_insert_for_auto(&pk_field, &mut post_insert);
-    post_insert.push(quote!(self.state.saved = true;));
+    let save_core = if is_auto(&pk_field) {
+        let pkident = pk_field.ident.clone().unwrap();
+        let values_no_pk: Vec<TokenStream2> = push_values(ast_struct, |f: &Field| f != &pk_field);
+        let save_cols = columns(ast_struct, |f| !is_auto(f) && f != &pk_field);
+        quote!(
+            // Since we expect our pk field to be invalid and to be created by the insert,
+            // we do a pure insert or update, no upsert allowed.
```

Comment:
not particular relevant at this stage, but might be worth using to clarify in this comment, but I believe we can do upserts with auto-increment (at least pg, mysql, havent checked sqlite), however this requires unique key constraints on other fields to tell the db engine when to update instead of inserting, and we dont have a way to do unique constraints yet.

Reply:
Good point